charmhelpers/contrib/sysctl/watermark_scale_factor.py

```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+# Copyright 2014-2015 Canonical Limited.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#  http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from charmhelpers.core.hookenv import (
+    log,
+    DEBUG,
+    ERROR,
+)
+
+import re
+
+WMARK_MAX = 1000
+WMARK_DEFAULT = 10
+MEMTOTAL_MIN_KB = 16777152
+MAX_PAGES = 2500000000
+
+
+def calculate_watermark_scale_factor():
+    """Calculates optimal vm.watermark_scale_factor value
+
+    :returns: watermark_scale_factor
+    :rtype: int
+    """
+
+    memtotal = get_memtotal()
+    normal_managed_pages = get_normal_managed_pages()
+
+    try:
+        wmark = min([watermark_scale_factor(memtotal, managed_pages)
+                     for managed_pages in normal_managed_pages])
+    except ValueError as e:
+        log("Failed to calculate watermark_scale_factor from normal managed pages: {}".format(normal_managed_pages), ERROR)
+        raise e
+
+    log("vm.watermark_scale_factor: {}".format(wmark), DEBUG)
+    return wmark
+
+
+def get_memtotal():
+    """Parse /proc/meminfo for memtotal value
+
+    :returns: memtotal
+    :rtype: int
+    """
+
+    memtotal = None
+    try:
+        with open('/proc/meminfo', 'r') as f:
+            for line in f.readlines():
+                if "MemTotal" in line:
+                    memtotal = int(re.search(r"\d+", line).group())
+                    break
+            else:
+                raise Exception("Could not find MemTotal")
+    except (Exception, OSError) as e:
+        log("Failed to parse /proc/meminfo in calculating watermark_scale_factor: {}".format(e), ERROR)
+        raise e
+
+    return memtotal
```

Comment:
Can you use https://github.com/juju/charm-helpers/blob/b5725ac546372e7d4004d15095f79cdd5e7da687/charmhelpers/core/host.py#L1116 instead? Having 2 mechanisms for calculating total memory might be harder to follow.

Reply:
I'm not adverse to it, although it does appear to return value in a different unit.

Reply:
done now